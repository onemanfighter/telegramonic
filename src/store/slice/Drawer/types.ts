export type DrawerData = {
  isOpen: boolean;
  title: string;
  children: React.ReactNode;
  footer: React.ReactNode;
  onModalClose: () => void;
};

export type DrawerState = {
  drawerData: DrawerData;
  openState: DrawerOpenState;
};

export enum DrawerOpenState {
  OPEN = 'OPEN',
  CLOSE = 'CLOSE',
}

export interface DrawerStateAction {
  openDrawer: (drawerData: DrawerData) => void;
  closeDrawer: () => void;
}

export type DrawerStateSlice = DrawerState & DrawerStateAction;
