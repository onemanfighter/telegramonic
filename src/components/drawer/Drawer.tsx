import {
  Drawer,
  DrawerBody,
  DrawerCloseButton,
  DrawerContent,
  DrawerFooter,
  DrawerHeader,
  DrawerOverlay,
} from '@chakra-ui/react';
import { DrawerProps } from './types';

const DrawerComponent = ({
  title,
  isOpen,
  children,
  footer,
  onModalClose,
}: DrawerProps) => {
  return (
    <Drawer isOpen={isOpen} placement="right" onClose={onModalClose}>
      <DrawerOverlay />
      <DrawerContent>
        <DrawerCloseButton />
        <DrawerHeader>{title}</DrawerHeader>

        <DrawerBody>
          {children}
          <DrawerFooter>{footer}</DrawerFooter>
        </DrawerBody>
      </DrawerContent>
    </Drawer>
  );
};

export default DrawerComponent;
