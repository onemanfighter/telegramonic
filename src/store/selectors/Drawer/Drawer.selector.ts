import { AppStoreState } from '@store';

const drawerSelector = (state: AppStoreState) => ({
  openState: state.Drawer.openState,
  drawerData: state.Drawer.drawerData,
  openDrawer: state.Drawer.openDrawer,
  closeDrawer: state.Drawer.closeDrawer,
});

export { drawerSelector };
