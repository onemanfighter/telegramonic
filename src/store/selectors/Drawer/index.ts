export { drawerSelector } from './Drawer.selector';
