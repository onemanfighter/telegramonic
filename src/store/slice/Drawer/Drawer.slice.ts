import { AppStoreSlice } from '../../appStore/types';
import { DrawerState, DrawerStateSlice, DrawerOpenState } from './types';

const initialAlertState: DrawerState = {
  drawerData: {
    isOpen: false,
    title: '',
    children: null,
    footer: null,
    onModalClose: () => {},
  },
  openState: DrawerOpenState.CLOSE,
};

const createDrawerSlice: AppStoreSlice<DrawerStateSlice> = (set) => ({
  ...initialAlertState,
  openDrawer: (modalData) => {
    set((state) => {
      state.Drawer.drawerData = modalData;
      state.Drawer.openState = DrawerOpenState.OPEN;
    });
  },
  closeDrawer: () => {
    set((state) => {
      state.Drawer.openState = DrawerOpenState.CLOSE;
    });

    setTimeout(() => {
      set((state) => {
        state.Drawer.drawerData = initialAlertState.drawerData;
      });
    }, 500 /* 0.5s */);
  },
});

export default createDrawerSlice;
