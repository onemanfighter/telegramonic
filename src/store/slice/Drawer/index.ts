export { default as createDrawerSlice } from './Drawer.slice';
export type { DrawerState, DrawerData, DrawerStateSlice } from './types';
export { DrawerOpenState } from './types';
