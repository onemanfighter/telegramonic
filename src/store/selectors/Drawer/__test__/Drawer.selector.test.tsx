import { AlertComponentState, DrawerState, appStore } from '@store';
import { act, renderHook } from '@testing-library/react';
import { drawerSelector } from '../Drawer.selector';

jest.useFakeTimers();

describe('Drawer selector', () => {
  const onDrawerCloseMock = jest.fn();
  const drawerData: DrawerState['drawerData'] = {
    title: 'title',
    isOpen: true,
    children: <div>children </div>,
    footer: <div>footer </div>,
    onModalClose: onDrawerCloseMock,
  };
  it('should return drawer selector state and actions', () => {
    const drawer = renderHook(() => appStore(drawerSelector)).result.current;

    expect(drawer).toMatchSnapshot();
  });

  it('should return alert state with initial value', () => {
    const { result } = renderHook(() => appStore(drawerSelector));

    expect(result.current.drawerData).toMatchSnapshot();
  });

  it('should return drawer state on drawer data set using openDrawer', () => {
    const { result } = renderHook(() => appStore(drawerSelector));

    act(() => {
      result.current.openDrawer(drawerData);
    });

    expect(result.current.drawerData).toMatchSnapshot();
  });

  it('should reset and close drawer state on drawer data set using closeDrawer', () => {
    const { result } = renderHook(() => appStore(drawerSelector));

    act(() => {
      result.current.openDrawer(drawerData);
    });
    expect(result.current.drawerData.isOpen).toBeTruthy();

    act(() => {
      result.current.closeDrawer();
      jest.runAllTimers();
    });

    expect(result.current.drawerData.title).toEqual('');
    expect(result.current.drawerData.isOpen).toBeFalsy();
  });
});
