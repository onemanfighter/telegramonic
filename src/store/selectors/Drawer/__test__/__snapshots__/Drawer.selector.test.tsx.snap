// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer selector should return drawer selector state and actions 1`] = `
{
  "closeDrawer": [Function],
  "drawerData": {
    "children": null,
    "footer": null,
    "isOpen": false,
    "onModalClose": [Function],
    "title": "",
  },
  "openDrawer": [Function],
  "openState": "CLOSE",
}
`;

exports[`Drawer selector should return drawer state on drawer data set using openDrawer 1`] = `
{
  "children": <div>
    children 
  </div>,
  "footer": <div>
    footer 
  </div>,
  "isOpen": true,
  "onModalClose": [MockFunction],
  "title": "title",
}
`;

exports[`Drawer selector should return drawer state with initial value 1`] = `
{
  "children": null,
  "footer": null,
  "isOpen": false,
  "onModalClose": [Function],
  "title": "",
}
`;
